Interactive debugger (CLI and DAP Server) for probe-rs

noppej opened the pull request:
This introduces some changes to probe-rs library as well as a new debugger component to allow interactive debugging on top of probe-rs. 

There is still lots of room for additions/completion, but I'm submitting it now because there is enough structure and functionality to demonstrate what is possible, and allow others to contribute in parallel. 

For now the binary is called `probe-rs-debugger`, and in conjunction with the [VSCode debug extension](https://github.com/probe-rs/vscode) , provides the following capabilities:

- [x] **Launch**: Automatcially launch probe-rs-debugger executable and connect to it, or ...
- [x] **Attach**: Use TCIP/IP port to connect to an existing probe-rs-debugger server
- [x] **Connect** to probe with probe-rs 
  - [x] Supports `connect-under-reset`
  - [ ] Only tested against STM32H745, using the Cortex-M7 core of this multi-core chip.
- [x] **Flash** the chip with your own binary. 
  - [x] Supports `reset-after-flashing`
  - [x] Supports `halt-after-reset`. This will allow you to set breakpoints in your main() function.
- [x] Set, clear, disable, enable **Breakpoints**
- [x] **Step Over** executing code
  - [x] Step Over works at 'instruction' granularity, so sometimes requires multiple steps per line of code
  - [ ] Stepping at 'line' level, Step Into, Step Out, does not work yet
- [x] **Variables View**
  - [x] View values of core **Registers**, and changes during code execution
    - [ ] TODO: Expand to show additional architecture registers
  - [x] View values of **Locals** variables, and changes during code execution.
    - [x] Shows datatypes and values for the following Rust datatypes.
      - [x] Base types, including &str
      - [x] Enumerations
      - [x] Structures
      - [x] Pointers
      - [x] Variants
    - [ ] TODO: Add support for all data types, such as Arrays, Unions, Generics, etc.
    - [ ] TODO: Filter the list of local variables to ONLY display variables that are in scope of the current register PC value. Currently ALL local variables are shown.
  - [ ] TODO: Expose **Static** and **Global** variables
- [x] **Call Stack View**
  - [x] Supports a single thread, for a single core of the chip, but will **allow selection of any frames** that are in the current thread
  - [ ] TODO: Support multiple threads
  - [ ] TODO: Support chips with multiple cores
- [ ] TODO: **Watch View** 
- [ ] TODO: Support for **ITM** or **RTT**

noppej commented:
Please note... Before we promote this branch, there are some fixes required in the README.md to remove references to the branch.

Tiwalun commented:
If you run `git rebase master` on your branch, and then force push it, this should get rid of the overwrites. Just ensure the master branch is updated before you run it.

Tiwalun commented on debugger/src/main.rs:
```diff
@@ -0,0 +1,184 @@
+}
+fn main() -> Result<()> {
+    // Initialize the logging backend.
+    pretty_env_logger::formatted_builder()
+        .target(pretty_env_logger::env_logger::Target::Stderr) // Log to Stderr, because the DebugAdapater, in 'Launch' mode, needs Stdin and Stdout to communicate with VSCode DAP Client
+        .init();
```
This seems to ignore the `RUST_LOG` environment variable. How can I configure which level is logged?

noppej replied:
Thanks for catching this. I am replacing pretty_env_logger with the underlying env_logger. It is leaner and I don't need all the features. In an ideal world, I'd like to find a way to direct all the RUST logs to the `debug_adapter::log_to_console()` method, but I need help with that ... because the debug_adapter is only initialized late in the startup. Any ideas?

Tiwalun replied:
You can have a look at https://github.com/probe-rs/probe-rs/blob/master/probe-rs-cli-util/src/logging.rs for ideas, @Yatekii should know more about it. Another solution can also be to just log into a file.

I'm also fine with improving this later, just being able to select the log level would be good enough for me. With the TCP server console logs work quite well.

noppej replied:
Excellent. I will add a TODO in the code. Once the PR has been promoted, I will systematically add Issues/PR's for individual TODO's.

Tiwalun commented on debugger/src/debugger.rs:
```diff
@@ -0,0 +1,1111 @@
+    /// The number associated with the debug probe to use. Use 'list' command to see available probes
+    #[structopt(long = "probe-index", default_value)]
+    #[serde(default)]
+    pub(crate) probe_index: usize,
```
Please use an `Option` here, otherwise this leads to problem when multiple probes are connected. Right now this defaults to 0, and seems to just open the first probe found, even when multiple probes are connected.

Yatekii replied:
Also, I would like to add, that we have moved to a better way of specifying the debug probe which is deterministic and not dependant on USB scan order for our other tools. E.g. https://github.com/probe-rs/cargo-flash/blob/master/src/main.rs#L74

noppej replied:
Ouch. Changing this impacts the VSCode extension also, but I can see that it is worth it.

Yatekii reviewed (commented):
Thanks a ton again for working on this!

I played a bit with the extension and a few things regarding UX caught my attention without yet looking at code:

- When trying to halt the core in vscode the DAP server crashes with `Probe(ProbeSpecific(CommandFailed(SwdApWdataError)))`. Did you experience that too? I feel like this is a probe-rs issue in general, but I think the DAP server should not crash.
- I feel like if I add and remove a few breakpoints, after some time it does not set them properly anymore. VSCode shows a hover info that the breakpoint is scheduled for placement (grey breakpoint).
- It seems to not get nice enough debug info from LLVM, but not much we can do here atm.
- It does not log much info and `RUST_LOG=debug` has no influence on logs.
- When I start the extension from within VSCode it prints no logs at all. When I start it externally and connect via TCP I see some logs in VSCode, but not really much useful stuff. Where does that difference come from.

After I fixed the macOS specific error @Tiwalun already pointed out, it went quite smooth, appart from the things you already pointed out in the PR description :)

I think we can very well coordinate efforts from on here to fix existing papercuts :)

noppej commented on debugger/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+schemafy = "^0.5"
+
+
+[profile.release]
+debug = true
+split-debuginfo = "unpacked"
+
+[profile.dev]
+split-debuginfo = "unpacked"
```
I agree. These are left over from early on, when I was holding back from changing the root Cargo.toml. I will check-in a fix soon.

noppej commented on debugger/src/debug_adapter.rs:
```diff
@@ -0,0 +1,1545 @@
+                            original_error,
+                        } => {
+                            // println!("temporary error ... retry: {}", os_error_number);
+                            if os_error_number == 35 {
```
Yep, tested it and it works. Thanks for the pointer :)

Tiwalun commented on .vscode/launch.json:
```diff
@@ -0,0 +1,95 @@
+                    "--bin=probe-rs-debugger",
+                    "--package=probe-rs-debugger",
+                    "--target",
+                    "x86_64-apple-darwin"
```
Does this also work without specifying the target? Otherwise it will only work on Macs.

noppej replied:
@Tiwalun ... yes, it works without specifying the target. I cleaned it up.

Tiwalun reviewed (commented):
Sorry for doing these comments in separate reviews, this should be the last of them.

noppej commented:
> Sorry for doing these comments in separate reviews, this should be the last of them.

No worries, this was a big PR, and I really appreciate the attention to detail in your review. I'm working my way through them, and will check-in an update soon. Or would you prefer smaller pushes?

noppej commented:
@Yatekii ... Some comments below for futher discussion,

> * When trying to halt the core in vscode the DAP server crashes with `Probe(ProbeSpecific(CommandFailed(SwdApWdataError)))`. Did you experience that too? I feel like this is a probe-rs issue in general, but I think the DAP server should not crash.
- No, I don't get that error on halt. If it persists for you we should try to trace it and catch it. In general, I find that the probe-rs Error structs catches errors behind a single string, and that obfuscates what the root cause of the error was. Perhaps we can think about a more consistent error handling for probe-rs components?
- Currently the DAP Server serves a single session, and then exits. If that session encounters an error with the Probe, it ends the DAP Client session - I do this because for Probe errors, it is hard to know how to reliably re-synch the DAP client to the new state. When the DAP Client session ends, it terminates the server. I've added a TODO to extend it to reconsider this behaviour, but it will need a larger implementation than I think will fit in this PR. Are you OK if we park it with a TODO for now?

> * I feel like if I add and remove a few breakpoints, after some time it does not set them properly anymore. VSCode shows a hover info that the breakpoint is scheduled for placement (grey breakpoint).
- Yes, I can reproduce this. It seems that the underlying probe-rs `clear_all_hw_breakpoints()` do not always succeed when it returns with an Ok, and then the debug_adapter tries to add more than the allowable number of hardware breakpoints. I am working on trying to pinpoint and fix this. 

> * It seems to not get nice enough debug info from LLVM, but not much we can do here atm.
- Agreed. What is the specific debug info you are looking for?

> * It does not log much info and `RUST_LOG=debug` has no influence on logs.
- RUST_LOG is now fixed (will be pushed soon). 
- I did not add a lot of rust log/warn/info statements,  because I chose to do logging via the DAP Client Debug Console, so that the user can see logs in the same UI where they are debugging. I'm open to discussion, and I have a TODO to find a way of integrating rust logs with DAP Client Debug Console logs. 

> * When I start the extension from within VSCode it prints no logs at all. When I start it externally and connect via TCP I see some logs in VSCode, but not really much useful stuff. Where does that difference come from.
- VSCode extenion logs are of marginal use (my humble opinion). I chose to implement user facing logs that shows up in the Debug Console for users who install the Extension the normal way. This is also the way gdb/openocd tends to work when you specify log levels in the launch.json. As always, I am open to discussion on that. Did you look at the "Debug" setting for "console_log_level" in the launch.json?
 
> After I fixed the macOS specific error @Tiwalun already pointed out, it went quite smooth, appart from the things you already pointed out in the PR description :)
> 
> I think we can very well coordinate efforts from on here to fix existing papercuts :)

noppej commented on probe-rs/src/architecture/arm/core/m4.rs:
```diff
@@ -478,6 +481,15 @@ impl<'probe> CoreInterface for M4<'probe> {
+        //First check if we stopped on a breakpoint, because this requires special handling before we can continue
+        let was_breakpoint =
+            if self.state.current_state == CoreStatus::Halted(HaltReason::Breakpoint) {
+                self.enable_breakpoints(false)?;
```
@Tiwalun ... I agree, this doesn't allow for the case of software breakpoints.  

I am not sure I understand `setting the program counter to the next instruction` ... would that not bypass the current instruction?

Tiwalun replied:
You're right, with HW breakpoints this would probably bypass the current instruction. 

In the case of SW breakpoints, this is exactly what you want, because the SW breakpoint is an instruction as well. 

This means that we  have to check if we hit a SW or HW breakpoint, and react differently. This is probably a bigger change, and should be done for all core types. I would suggest we open an issue for this, and handle it separately, what do you think? If we handle it separately, this can be merged as is, it's already an improvemen over the current statet.

noppej replied:
@Tiwalun ... Cool, I agree. The debugger will eventually need to set its own software breakpoints if we want to implement stepping at line level, etc. , so handling SW breakpoints is a good thing to do. I will create an Issue on the master branch for this.

Tiwalun replied:
Great, thanks.

noppej commented on probe-rs/src/debug/variable.rs:
```diff
@@ -1,10 +1,319 @@
+    }
+}
+
+///Define the role that a variable plays in a Variant relationship
```
@Tiwalun, I've updated it as follows ... does this help?
```
///Define the role that a variable plays in a Variant relationship. See section '5.7.10 Variant Entries' of the DWARF 5 specification
#[derive(Debug, Clone, PartialEq)]
pub enum VariantRole {
    ///A (parent) Variable that can have any number of Variant's as it's value
    VariantPart(u64),
    ///A (child) Variable that defines one of many possible types to hold the current value of a VariantPart. 
    Variant(u64),
    ///This variable doesn't play a role in a Variant relationship
    NonVariant,
}
```

Tiwalun replied:
Yes, that's great, thanks!

noppej commented on probe-rs/src/debug/variable.rs:
```diff
@@ -1,10 +1,319 @@
 
-#[derive(Debug, Default)]
+#[derive(Debug, Clone, PartialEq)]
+pub enum VariableKind {
```
@Tiwalun ... does this help?
```
///VariableKind is a tag used to differentiate the nature of a variable. The DAP protocol requires a differentiation between 'Named' and 'Indexed'. We are adding 'Referenced', because those require unique handling when decoding the value during runtime
#[derive(Debug, Clone, PartialEq)]
pub enum VariableKind {
    ///An Indexed variable (bound to an ordinal position), such as the sequenced members of an Array or Vector
    Indexed,
    ///A variable that is identified by it's name, and is not bound to a specific ordinal position.
    Named,
    ///A variable that is the target of a pointer variable
    Referenced,
    ///This should never be the final value for a Variable
    Undefined,
}
```

Tiwalun replied:
Looks good, thanks!

noppej commented on debugger/src/debugger.rs:
```diff
@@ -0,0 +1,1111 @@
+                .debugger_options
+                .program_binary
+                .clone()
+                .unwrap()
```
@Tiwalun, Thanks. I fixed two instances of that issue.

noppej commented on debugger/src/debug_adapter.rs:
```diff
@@ -0,0 +1,1545 @@
+}
+
+pub(crate) trait DapStatus {
+    fn short_long_status(&self) -> (&str, &str);
```
@Tiwalun ... another good catch. Thank you :)

noppej commented on debugger/src/debugger.rs:
```diff
@@ -0,0 +1,1111 @@
+    }
+}
+
+/**
```
@Tiwalun ... I've made the change to comply with consistency. Although I would like to understand the reluctance to use the /** */ feature? I personally find it makes it easier to create and edit nicely formatted markdown comments.

Tiwalun replied:
I'm very much used to seeing only `///` for Rust comments, so I was a bit confused when I saw this. I didn't know that this style of comments is even possible for doc comments.

Is there any difference in formatting between these comments? I would have expected that they both work the same in regards to that.

noppej replied:
@Tiwalun, The way I read the rust docs, the `///` is for lines of comments, where `/** */` is for blocks. I don't think it matters too much the way you prefer it, but I just like the block style comments.

```
OUTER_LINE_DOC :
   /// (~/ ~[\n IsolatedCR]*)?

OUTER_BLOCK_DOC :
   /** (~* | BlockCommentOrDoc ) (BlockCommentOrDoc | ~[*/ IsolatedCR])* */
```

noppej commented on debugger/src/debugger.rs:
```diff
@@ -0,0 +1,1111 @@
+    target_probe.select_protocol(
+        debugger_options
+            .protocol
+            .or(Some(WireProtocol::Swd))
+            .unwrap(),
```
@Tiwalun ... done :)

noppej commented on debugger/src/debugger.rs:
```diff
@@ -0,0 +1,1111 @@
+    } else {
+        target_probe.attach(target_selector).map_err(|err| {
+            anyhow!(
+                "Error creating capstone: {:?}.\nTry the --connect-under-reset option",
```
@Tiwalun, I'm impressed ... that was a subtle, but valuable, catch. Thank you :)

noppej commented on debugger/src/debugger.rs:
```diff
@@ -0,0 +1,1111 @@
+}
+
+//TODO: When we implement multi-threaded server, this static handling needs work
+pub static mut LAST_KNOWN_STATUS: CoreStatus = CoreStatus::Unknown;
```
@Tiwalun ... moving it to the struct is much cleaner. Done :)

noppej commented on probe-rs/src/debug/variable.rs:
```diff
@@ -1,10 +1,319 @@
+            name: String::new(),
+            value: String::new(),
+            file: String::new(),
+            line: u64::MAX, //A default value that tells us it hasn't been processed yet.
```
@Tiwalun, The u64::MAX is not really needed. It is never used, and it doesn't have to be an `Option` either. It is OK for it to have the default value of 0 when there is no line number associated. I've fixed it as such. Are you OK with that approach?

Tiwalun replied:
Looks good to me.

noppej commented on probe-rs/src/debug/variable.rs:
```diff
@@ -1,10 +1,319 @@
-    pub value: u64,
-    pub typ: Type,
+    pub type_name: String,
+    pub location: u64,
```
@Tiwalun, I've renamed and documented it as follows ...
```
///The starting location/address in memory where this Variable's value is stored.
pub memory_location: u64,
```

Better?

Tiwalun replied:
Yes, completely clear now.

Tiwalun commented on .vscode/launch.json:
```diff
@@ -0,0 +1,90 @@
+                    "build",
+                    "--bin=probe-rs-debugger",
+                    "--package=probe-rs-debugger",
+                    "--target",
+                    "x86_64-apple-darwin"
```
I think you missed one `--target` here.

Tiwalun reviewed (approved):
Just the one target in `.vscode/launch.json` left to remove, otherwise this looks good to me!

Thanks again for your work!

Yatekii reviewed (approved):
> * No, I don't get that error on halt. If it persists for you we should try to trace it and catch it. In general, I find that the probe-rs Error structs catches errors behind a single string, and that obfuscates what the root cause of the error was. Perhaps we can think about a more consistent error handling for probe-rs components?

Yes, at least on my STM32H747 it's persistent. I'll try with another chip/probe.
Errors are a long standing issue. It is hard to get them right. We want a maximum amount of information with a minimal amount of clutter and indirection. Sadly that's not easy. We are open to suggestions. @Tiwalun and I made a lot of efforts to improve errors but it's a really tall order.

> * Currently the DAP Server serves a single session, and then exits. If that session encounters an error with the Probe, it ends the DAP Client session - I do this because for Probe errors, it is hard to know how to reliably re-synch the DAP client to the new state. When the DAP Client session ends, it terminates the server. I've added a TODO to extend it to reconsider this behaviour, but it will need a larger implementation than I think will fit in this PR. Are you OK if we park it with a TODO for now?

Yes, this is perfectly fine with me, thanks!

> * Yes, I can reproduce this. It seems that the underlying probe-rs `clear_all_hw_breakpoints()` do not always succeed when it returns with an Ok, and then the debug_adapter tries to add more than the allowable number of hardware breakpoints. I am working on trying to pinpoint and fix this.

Ok, I'll conduct some experiments here too then!

> * Agreed. What is the specific debug info you are looking for?

I tried to set a few breakpoints on pretty much trival lines of code, like some variable assignments. And it said it couldn't find that code section. Which means there was no debug info for this line. Maybe we also do not do everything properly with the line determination algorithm.

> * RUST_LOG is now fixed (will be pushed soon).
> * I did not add a lot of rust log/warn/info statements,  because I chose to do logging via the DAP Client Debug Console, so that the user can see logs in the same UI where they are debugging. I'm open to discussion, and I have a TODO to find a way of integrating rust logs with DAP Client Debug Console logs.
>
> * VSCode extenion logs are of marginal use (my humble opinion). I chose to implement user facing logs that shows up in the Debug Console for users who install the Extension the normal way. This is also the way gdb/openocd tends to work when you specify log levels in the launch.json. As always, I am open to discussion on that. Did you look at the "Debug" setting for "console_log_level" in the launch.json?

This sounds nice to me, thanks! Sometimes the DAP messages are a bit spammy, but at least they contain all the info :) At first we struggled to enable it, but we just oversaw the enable in the launch.json. It should be quite obvious, but we overlooked it somehow :)

All in all, this looks like very nice work, thank you :)

One minor nit that we can fix later, generally, comment commands should be followed by a space (like `/// x` and not `///x`). I think I saw this quite a lot. Just so you know. This is no blocker for this PR.

I think @Tiwalun already pointed out a lot of other stuff which is enough for me. We can improve & add more code over time. I think it is good to merge after the last minor fix in the `launch.json` example :)

noppej commented:
@Yatekii ... Thanks. I appreciate all the positive feedback and help getting this PR cleaned up.  Some final notes before I check this in ...

- I've done some digging on the breakpoints, and think I've got it figured out ... will fix it after this PR is through.
- I ran a global regex search and replace, and added a space between /// and //, and their following text ... just because it's so easy to fix.

Yatekii commented:
Thanks a lot!

bors r+
